Migrate keyboards using uGFX to LED_MATRIX

firetech opened the pull request:
Migrate Input Club keyboards (Ergodox Infinity and Whitefox) away from uGFX for keyboard backlight.

## Description

Ergodox Infinity and Whitefox use uGFX to control their keyboard backlight. Because of licensing issues, it's desirable to move away from uGFX (see #5268). The chip in question (IS31FL3731C) already had a working driver for use with LED Matrix (drivers/issi/is31fl3731-simple.c).

Since uGFX is still compiled for the LCD display in Ergodox Infinity,  and it caused some naming conflicts, I had to rename the `point_t` type of LED/RGB Matrix. `led_point_t` seemed fitting. (This is in conflict with two lines in #12651, but is an easy fix.)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented:
I suspect that it will need more changes to get this working properly (read: 100%), but it looks like a good start, at least!

tzarc commented:
Given that it's changing a bunch of things at once, it'll need to go to `develop` instead of `master`.

firetech commented:
> I suspect that it will need more changes to get this working properly (read: 100%), but it looks like a good start, at least!

Definitely. For now, I'm waiting for the i2c support to be merged before doing more work on this.

drashna commented:
Looks like there is an issue with the whitefox board: 
```
Making whitefox with keymap default                                                                    [ERRORS]
lib/ugfx/src/gdisp/gdisp.c: In function '_gdispInit':
lib/ugfx/src/gdisp/gdisp.c:566:4: error: useless type name in empty declaration [-Werror]
    extern GDISPVMTLIST        GDISP_DRIVER_LIST;
    ^~~~~~
cc1: all warnings being treated as errors
 | 
 | 
 | 
make[1]: *** [tmk_core/rules.mk:386: .build/obj_whitefox/lib/ugfx/src/gdisp/gdisp.o] Error 1
make[1]: *** Waiting for unfinished jobs....
```

firetech commented:
> Looks like there is an issue with the whitefox board:

Yeah, I haven't touched the Whitefox yet, but I have changed some common code in quantum visualizer, so it's entirely expected that it doesn't compile yet. ;)

Making this posting as a PR might have been jumping the gun a bit, it might have been better to make a normal issue and just refer to my branch...

fauxpark commented:
The I2C PR has been merged now, so this should be good to progress.
Additionally, I just discovered the Infinity60 (also by Input:Club) uses the 3731C too, but has some custom controller code instead of uGFX.

skullydazed commented:
@firetech do you think you'll be working on this again?

tzarc commented:
I... have no idea why I didn't review this previously.
I've sorted out conflicts as best I could -- would be great to get things validated.
Would prefer to get the `#TODO: Remove` stuff removed for the PR merge, though.

fauxpark commented:
I'm happy to report that I have LED matrix more or less caught up with RGB matrix and ported over a couple animations to test. I just need to break down the giant diff into more digestible PRs.

Should make this PR much more useful :)

firetech commented:
> @firetech do you think you'll be working on this again?

Maybe, but probably not in the near future. Life has been lacking free time and motivation lately... If someone else would like to continue the work, be my guest.

firetech commented:
This is almost working properly now. The LED positions in `g_led_config` are all zero (I'll use the data in info.json to calculate proper positions at some point), and I'm having some issues with `led_matrix_eeconfig.flags` misbehaving (see discussion in #12580).

However, my keyboard is behaving slightly random when it comes to waking up after suspend (sometimes not waking at all, sometimes waking but not calling `suspend_wakeup_init()`, and sometimes working properly), but I think it's the same on master currently (I haven't updated my daily driver firmware in a while)...

firetech commented:
Forgot to mention that the newest code no longer breaks the visualizer since LED Matrix has been decoupled from the Backlight system. Both old and new models of LED control are therefore (hopefully) supported (but will probably break if you try to run both at the same time, of course).

This also should make it possible to merge this without porting Whitefox.

firetech commented:
I think this is (finally!) ready for review. 😄

fauxpark commented:
#12845 may help with that.

firetech commented:
Since #12651 was merged, I'll have to do a small update of this to make it compileable again. Don't merge this yet!

fauxpark commented:
@firetech I've opened the above linked PR just to cut down on this PR's scope. The SDB stuff I want to handle differently, by simply specifying a pin in config.h and the driver will do the needful (including possibly deasserting it when the user toggles the LED/RGB matrix off etc to save power).

firetech commented:
> @firetech I've opened the above linked PR just to cut down on this PR's scope. The SDB stuff I want to handle differently, by simply specifying a pin in config.h and the driver will do the needful (including possibly deasserting it when the user toggles the LED/RGB matrix off etc to save power).

Sounds reasonable. I mostly made this quick fix since the driver didn't handle it at all previously (when I started writing the patch last summer, no board using LED Matrix and IS31FL3731 seemed to have SDB connected to an MCU pin).

fauxpark commented:
> @fauxpark Do you intend to work on that, or what is the plan?

Yup, but it might be a short while as technically develop is now locked to new PRs (apart from urgent bugfixes, of course).

firetech commented:
Updated to be based on #12864 (not including it, but can't be built without it), and moved the IS31FL3731 SDB stuff to `keyboard_pre_init_kb()` (`matrix_init_kb()` is called after `led_matrix_init()`, so that didn't work), in anticipation of an improved, common way of setting SDB.

fauxpark commented on keyboards/whitefox/whitefox.c:
```diff
@@ -16,3 +16,96 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #include "whitefox.h"
+
+#ifdef LED_MATRIX_ENABLE
+#    include "led_matrix.h"
+#    include "is31fl3731-simple.h"
+
```
```suggestion
```
These includes aren't needed anymore.

fauxpark commented on keyboards/whitefox/whitefox.c:
```diff
@@ -16,3 +16,96 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+void keyboard_pre_init_kb(void) {
+#ifdef LED_MATRIX_ENABLE
+    // Turn on LED controller
+    palSetPadMode(GPIOB, 16, PAL_MODE_OUTPUT_PUSHPULL);
+    palSetPad(GPIOB, 16);
```
```suggestion
    setPinOutput(B16);
    writePinHigh(B16);
```
Does this not work?

firetech replied:
Apparently, yes. The `palSet...` lines were copied from the old uGFX/visualizer code, I had no idea the cleaner syntax existed. 🙂

fauxpark replied:
I recall there are some naming conflicts (yet again...) with Kinetis in ChibiOS as some registers have the same names as certain QMK-style pins, so you cannot reference them in `MATRIX_COL/ROW_PINS` in config.h with `#include "config_common.h"` present, otherwise the compiler complains.

fauxpark commented on keyboards/ergodox_infinity/ergodox_infinity.c:
```diff
@@ -226,3 +295,117 @@ const keypos_t PROGMEM hand_swap_config[MATRIX_ROWS][MATRIX_COLS] = {
 #endif
+
+#ifdef LED_MATRIX_ENABLE
+#    include "is31fl3731-simple.h"
+
```
```suggestion
```

fauxpark commented on keyboards/ergodox_infinity/ergodox_infinity.c:
```diff
@@ -1,20 +1,46 @@
+#    include "serial_link/protocol/transport.h"
+
+#    ifdef LED_MATRIX_ENABLE
+#        include "led_matrix.h"
```
```suggestion
```

firetech commented:
Fixed the changes requested by @fauxpark, and removed some code made unnecessary by #12878.

fauxpark commented on keyboards/whitefox/whitefox.c:
```diff
@@ -16,3 +16,79 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+    // put your keyboard start-up code here
+    // runs once when the firmware starts up
+
+#ifdef LED_MATRIX_ENABLE
+    /*
+     * Since K20x is stuck with a 32 byte EEPROM (see tmk_core/common/chibios/eeprom_teensy.c),
+     * and neither led_matrix_eeconfig.speed or .flags fit in this boundary, just force their values to default on boot.
+     */
+#    if !defined(LED_MATRIX_STARTUP_SPD)
+#        define LED_MATRIX_STARTUP_SPD UINT8_MAX / 2
+#    endif
+    led_matrix_set_speed(LED_MATRIX_STARTUP_SPD),
+    led_matrix_set_flags(LED_FLAG_ALL);
+#endif
```
Ideally, I would like to see if something like #12947 for the K20x fixes this, before merging.

firetech replied:
As far as my research has lead me to understand, you basically need to wipe the entire MCU and reflash the bootloader to change the "EEPROM" size. 🙁

fauxpark replied:
Then I'm not sure how else this can be solved if we want VIA support (and room for future core settings)... perhaps we should just fall back to the transient EEPROM implementation, that will eliminate the need for this code at least.

firetech replied:
I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU. When I tried increasing `EEPROM_SIZE` locally, I guess my board therefore just crashed due to the firmware configuration not matching the hardware configuration.

The documentation is a bit fuzzy on the subject, but NXP Support claimed in their forum that the relevant flash and registers need to be in "an erased state" before running the PGMPART command. I'm a bit hesitant to try since I don't want to risk bricking my main keyboard, but it might be possible to try skipping the [`FTFL->FCNFG & FTFL_FCNFG_RAMRDY` if statement](https://github.com/qmk/qmk_firmware/blob/master/tmk_core/common/chibios/eeprom_teensy.c#L99) and that way force the PGMPART command to be run. Due to what NXP Support said in the forums, my guess is that it won't work.

fauxpark replied:
> I've done some more investigation and realized that the partitioning code in eeprom_teensy.c isn't actually run on an already partitioned MCU.

As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently). This code seems to be mainly intended for actual Teensys, which have an external chip doing the bootloading, and according to the PJRC forum thread mentioned in the other PR, it sends the `SETRAM` command to turn the EEPROM functionality off. It also likely does a mass erase first.

And yeah, the manual says:
> After clearing CCIF to launch the Program Partition command, the flash memory module first verifies that the EEPROM Data Size Code and FlexNVM Partition Code in the data flash IFR are erased.

firetech replied:
> As far as I can tell, the original Kiibohd firmware doesn't do any partitioning (so there is no persistent storage, apparently).

Hmm, so it might have been QMK that set my keyboard to 32 bytes of EEPROM, then?

firetech replied:
Sooo, I finally mustered up the courage to try and force the partitioning command to run, and, as expected, it had no effect whatsoever (I did make sure the code actually ran, since `eeprom_initialize` isn't even called if the EEPROM is ready before the first read or write). Reading the manual, it seems like the only way to clear the data flash IFR (after which partitioning should be allowed again) is the "clear all blocks" command, which undoubtedly would remove the bootloader. Not really a safe move...

The transient EEPROM driver might be the way to go for non-Teensy K20x boards... 😢 

I guess crazy users (like me 😛) who _really_ want to use EEPROM could theoretically implement a custom (hackish) keymap level EEPROM driver that reshuffles the eeconfig addresses to more effectively use the 32 available bytes. \*evil laugh*

fauxpark replied:
At least the bootloaders for all Input:Club boards can be found here: https://github.com/kiibohd/controller/releases
`ergodox.v0.5.7.bootloader.bin`

All you need is a J-Link, I suppose.

firetech replied:
How should we go ahead with this? I feel like changing the boards to the transient EEPROM driver might be a bit out of scope for this PR, but without either that or these lines of code, LED Matrix is effectively dead after the second boot (due to `flags` being 0)...

fauxpark replied:
I'm fine to merge this as is, now that I know a bit more about the problem.

Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.
Failing that, perhaps we can adapt the STM32 emulated eeprom?

firetech replied:
> Since this is technically a bug in QMK, it should get a proper fix. I would kinda like to fix the code in a separate, breaking changes PR - it would mean users will need to wipe their boards, of course. But, it should only need to be done once.

I have some doubts about how many users would be willing (and able) to wipe their boards like that...

If you just increase the default `EEPROM_SIZE` without changing anything else, boards initialized with another value will fail to boot. At least that's how mine behave.

firetech replied:
Hmm. It might even be possible to switch the FlexRAM function over from EEPROM to RAM when an incompatible EESIZE is detected in `SIM->FCFG1`, in effect making the EEPROM driver work using the same code, but behaving transient. I'm gonna do some experimenting. 👨‍🔬

firetech replied:
[Working prototype](https://github.com/firetech/qmk_firmware/commit/kinetis_flexram_fallback)! 😄

I also realized that the "crash" during boot after increasing `EEPROM_SIZE` was just an infinite loop in `flexram_wait()` due to the `EEERDY` flag never being set after an out-of-bounds write. Just making sure the FlexRAM isn't written to outside the partitioned EEPROM size is enough to make the board boot properly. In other words, the entire `if (eeprom_size < EEPROM_SIZE) { ... }` block _can_ be removed (but it might be desired).

If you feel that this would be a valid solution to the problem, I can make a PR of this after #12947 has been merged (would conflict heavily otherwise).

fauxpark replied:
An issue detailing all of these findings may be a good idea, rather than having it all in this tangentially-related PR discussion.

firetech replied:
Good point... I'll get one started.

firetech replied:
#13164

fauxpark commented on keyboards/whitefox/config.h:
```diff
@@ -62,6 +62,20 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#define LED_DRIVER_ADDR_1 0x74
+#define LED_DRIVER_COUNT 1
+#define DRIVER_LED_TOTAL 71
+#define LED_DISABLE_WHEN_USB_SUSPENDED true
```
```suggestion
#define LED_DISABLE_WHEN_USB_SUSPENDED
```
#13146

fauxpark commented on keyboards/ergodox_infinity/config.h:
```diff
@@ -62,6 +62,21 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#define LED_DRIVER_COUNT 1
+#define DRIVER_LED_TOTAL 76
+#define LED_MATRIX_SPLIT { 38, 38 }
+#define LED_DISABLE_WHEN_USB_SUSPENDED true
```
```suggestion
#define LED_DISABLE_WHEN_USB_SUSPENDED
```

fauxpark reviewed (changes requested):
One last thing, then I think I can merge this :)

firetech commented:
Fixed.

Was considering adding a check of `EEPROM_SIZE` to the eeprom limit hack in matrix_init_kb discussed above, but realized that `EEPROM_SIZE` probably isn't defined in that context...

fauxpark commented:
Thanks for all your work!
